### PR TITLE
feat(router): add a router plugin

### PR DESCRIPTION
**Pull request by Tieske**
This builds on top of #130 

Adds a router plugin.

### CODE REVIEW

File: rockspecs/pegasus-dev-1.rockspec [hunk 1]
```diff
@@ -38,5 +38,7 @@ build = {
     ['pegasus.response'] = 'src/pegasus/response.lua',
     ['pegasus.compress'] = 'src/pegasus/compress.lua',
     ['pegasus.plugins.compress'] = 'src/pegasus/plugins/compress.lua',
+    ['pegasus.plugins.downloads'] = 'src/pegasus/plugins/downloads.lua',
+    ['pegasus.plugins.tls'] = 'src/pegasus/plugins/tls.lua',
```
Comment (EvandroLG) [inline rockspecs/pegasus-dev-1.rockspec:42]: We should add the router plugin here as well

**Comment by wmealing** [comment]
I got tricked for a while, as I had forgotten that the string paths were regular expressions.  This post is left for other users who may wonder why pegas is not serving matching the expected path.

Any of the "magic characters"

`    ( ) . % + - * ? [ ^ $
`
Need to be escaped with %, for example the route "/catch-fish"

```
local routes = {

   ["/catch%-fish"] = {
      GET = function(req, resp)
         resp:statusCode(200)
         resp:addHeader("Content-Type","text/html")
         resp:write("catch")
      end,
   },
```

I do not consider this a bug int the code, just likely a bug in my comprehension.

**Review by wmealing — APPROVED** [review]
I'm relatively new with lua, so I can't speak for how it should look, however it does seem to work well.  

I tested the 'routing' part of this patchset,  it works well.

I got tripped up because I thought the path string was.. a string match originally,but after realising it was a regular expression,  it was clear.  Its possible that maybe this should be in the comments/documentation but its not a dealbreaker.

**Comment by Tieske** [comment]
@wmealing I think that's actually a bug, we should (for simplicity sake) escape the path before matching. Such that a path `"/my-cool-path/{parameter}"` can be used without manually escaping the "`-`" characters.

Other things:
- I think the shortcut on the path level (passing a function instead of a table) should go. Require users to be explicit and add a wildcard method handler "*" instead.
- The examples currently refer to a JSON lib, which was removed, so they must be updated.
- The last segment should be configurable as `"{parameter+}"` where the "+" means matching more than 1 segment (normally parameters are limited to 1 segment)
- Test that multiple router plugins can be added, such that multiple independent API components can be served from their own plugin. For example:
  - first plugin on: GET "/my/app"
  - second plugin on: POST "/my/app"
- Test that we unescape the path before matching, eg incoming `"/my%20path"` should match a configured router path `"/my path"` (also applies to the path-prefix on router level)
- Per router a list of hosts to match (currently only matches on a path prefix)

Any input appreciated.

**Review by EvandroLG — APPROVED** [review]
LGTM, thanks @Tieske

**Pull request merged**